src/core/inductive.ml

```diff
@@ -0,0 +1,255 @@
+  let p_iden = _P_Iden (Pos.none ([], rec_sym.sym_name)) true in
+  let p_patt = create_patt (Pos.none "p") in
+  let head   = _P_Appl p_iden p_patt in
+  let arg : sym list -> p_term = fun l  ->
+    let rec aux : sym list -> p_term -> p_term = fun l acc ->
+      match l with
+      | []   -> acc
+      | t::q ->
+          let p_patt = create_patt (Pos.none ("p" ^ t.sym_name)) in
+          aux q (_P_Appl acc p_patt)
+    in
+    aux l head
+  in
+  let common_head = arg cons_list in
+
+  (* STEP 2: Create each p_rule according to a constructor *)
+  (* [create_p_rule cons_sym] creates a p_rule according to a constructor
+     [cons_sym]. *)
+  let create_p_rule : sym -> p_rule = fun cons_sym ->
+    let pt       = Pos.none ("p" ^ cons_sym.sym_name) in (* RHS *)
+    let p_patt   = create_patt pt in
+    let qident_t = Pos.none ([], cons_sym.sym_name) in (* LHS *)
+    let t_ident  = _P_Iden qident_t true in
+    let rec aux : p_term list -> term -> p_term list -> int -> p_rule =
+      fun arg_list t hyp_rec_list i ->
+      match Basics.get_args t with
+      | (Symb(s), ts) ->
+          if s == ind_sym then
+            let (lhs_end, rhs_x_head) =
+              let tmp = fapp (List.rev arg_list) in
+              tmp t_ident, tmp p_patt
+            in
+            let ts = List.map type_to_pattern ts in
+            (*let len = List.length l in
+            let arg_type = app common_head (create_arg len) in*)
+            let arg_type = app common_head ts in
+            let lhs_x = _P_Appl arg_type lhs_end in
+            let rhs_x = app rhs_x_head hyp_rec_list in
+            if !log_enabled then
+              log_ind "The rule [%a] --> %a"
+                Pretty.pp_p_term lhs_x Pretty.pp_p_term rhs_x;
+            Pos.none (lhs_x, rhs_x)
+            else assert false (* See the function named "principle" *)
+      | (Prod(a, b), _) ->
+          let (_,b) = Basics.unbind_name b in
+          begin
+            match Basics.get_args a with
+            | (Symb(s), ts) ->
+                let arg_patt =
+                  create_patt (Pos.none ("x" ^ (string_of_int i)))
+                in
+                if s == ind_sym then
+                  let ts = List.map type_to_pattern ts in
+                  (*let len = List.length l in
+                  let arg_type = app common_head (create_arg len) in*)
+                  let arg_type = app common_head ts in
+                  let hyp_rec_x = _P_Appl arg_type arg_patt in
+                  let new_hyp_rec_x = hyp_rec_x::hyp_rec_list in
+                  aux (arg_patt::arg_list) b new_hyp_rec_x (i+1)
+                else aux (arg_patt::arg_list) b hyp_rec_list  (i+1)
+            | _ -> assert false (* See the function named "principle" *)
+          end
+      | _ -> assert false (* See the function named "principle" *)
+    in
+    aux [] (!(cons_sym.sym_type)) [] 0
+  in
+
+  (* STEP 3: Build all the p_rules *)
+  List.rev (List.map create_p_rule cons_list)
```

Comment:
no need for a List.rev: the order of rules is not important (btw List.rev_map is more efficient than List.rev List.map)

Reply:
use List.mapi to associate an integer to each constructor